### PR TITLE
Updates, more structures, and more REST covereage

### PR DESCRIPTION
**Pull request by CadetMercury615**
Focused this PR on expanding rest coverage rather than low-level features. Added more structures to interact with the API and added more methods within those structures. 

Channels were a mess on my end and thus I will make another PR once I fix them.

### CODE REVIEW

**Comment by Bavfalcon9** [comment]
Updating your fork will auto update this pull request

File: src/structures/Message.ts [hunk 1]
```diff
@@ -16,14 +17,16 @@ class Message extends Base {
...
+    public guild: Guild;
     public content: string;
     public channel_id: string;
     public channel: TextChannel;
     public author: User;
     public attachments: any[]; // Leave as any for now
     public args: string[];
 
-    constructor(data: any, channel?: TextChannel) {
+    constructor(data: any, channel?: TextChannel, guild?: Guild) {
```
Comment (Bavfalcon9) [inline src/structures/Message.ts:29]: You're assuming guild is automatically cached, this isn't the case and needs to be removed.

File: src/rest/RequestManager.ts [hunk 2]
```diff
@@ -114,5 +117,53 @@ class RequestManager {
                return await response.json();
           }
      }
+
```
Comment (Bavfalcon9) [inline src/rest/RequestManager.ts:120]: I'll probably be changing this a lot (this is a note for myself)

File: src/structures/Permission.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import Base from "./Base.ts";
```
Comment (Bavfalcon9) [inline src/structures/Permission.ts:1]: I have better implementation planned for this, but it works for now.

File: src/structures/Permission.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import { Permissions } from "./Permissions.ts";
+
+class Permission extends Base {
+    public allow: number;
+    public deny: number;
+    public _json: object = {};
+    constructor(allow: number, deny: number = 0) {
+        super('')
```
Comment (Bavfalcon9) [inline src/structures/Permission.ts:9]: Inconsistencies with semicolon.

File: src/structures/guild/Emoji.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    public roles: string[];
+    public requireColons: boolean;
+    public name: string;
+    public managed: boolean;
+    public available: boolean;
+    public animated: boolean;
+
+    constructor(data: any) {
+        super(data.id)
+        this.roles = data.roles
```
Comment (Bavfalcon9) [inline src/structures/guild/Emoji.ts:13]: Again, with semicolons.

File: src/structures/guild/Guild.ts [hunk 6]
```diff
@@ -70,16 +77,61 @@ class Guild extends Base {
...
+   }
+
+    public get paritial(): boolean { return false }
+
+    /**
+     * Channel Methods
+     */
+    public getTextChannels(): GuildChannel[] { return this.channels.filter((c: GuildChannel) => c.type === 0) }
+    public getVoiceChannels(): GuildChannel[] { return this.channels.filter((c: GuildChannel) => c.type === 2) }
+    public getCategoryChannels(): GuildChannel[] { return this.channels.filter((c: GuildChannel) => c.type === 4) }
```
Comment (Bavfalcon9) [inline src/structures/guild/Guild.ts:109]: This doesn't need to be built into the api.

File: src/structures/guild/Guild.ts [hunk 7]
```diff
@@ -70,16 +77,61 @@ class Guild extends Base {
...
+        return this.roles.find(r => r.id === roleID) || null
     }
+
+    public async getAdminRoles(): Promise<Role[]> {
+        return this.roles.filter(r => r.permissions === 104320577) as Role[];
+    }
+
+    public async getModRoles(): Promise<Role[]> {
+        return this.roles.filter(r => r.permissions === 49675847) as Role[];
+    }
```
Comment (Bavfalcon9) [inline src/structures/guild/Guild.ts:125]: These should not be built in to the api, i recommend making a utility for this.

File: src/structures/guild/GuildMember.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import Base from "../Base.ts";
+import User from "../User.ts";
+import Guild from "./Guild.ts";
+import Role from "./Role.ts";
+
+class GuildMember extends Base {
+    public user: User;
```
Comment (Bavfalcon9) [inline src/structures/guild/GuildMember.ts:7]: It's possible for this to be undefined.

File: src/structures/guild/Role.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+class Role extends Base {
+    public name: string;
+    public permissions: any;
+    public color: string;
+    public hoist: boolean;
+    public mentionable: boolean;
+
+    constructor(data: any) {
+        super(data.id)
```
Comment (Bavfalcon9) [inline src/structures/guild/Role.ts:11]: Inconsistencies with semicolons.

File: src/structures/guild/Guild.ts (hunk 7 above)
Comment (Bavfalcon9) [reply src/structures/guild/Guild.ts:125]: This also doesn't need to be async 🤔

**Comment by Bavfalcon9** [comment]
`GuildChannel` is missing.

**Comment by CadetMercury615** [comment]
Is `GuildChannel` still being implemented, or are we just dropping it to just have channels extend from `Channel`

**Comment by CadetMercury615** [comment]
If this looks good, I can update the rest of channels structures and roles for this

**Pull request closed without merging**